src/lib/core/mp_datetime.c

```diff
@@ -0,0 +1,194 @@
+  |0xC7|len (uint8)| 4 | seconds (int) | tzoffset (int)  | tzindex (uint) | nsec (int) |
+  +----+-----------+---+====~~~~~~~====+....~~~~~~~~~....+----~~~~~~~~----+....~~~~....+
+
+  MessagePack extension MP_EXT (0xC7), after 1-byte length, contains:
+
+  - signed integer seconds part (required). Depending on the value of
+    seconds it may be from 1 to 8 bytes positive or negative integer number;
+
+  - [optional] fraction time in nanoseconds as unsigned integer.
+    If this value is 0 then it's not saved (unless there is offset field,
+    as below);
+
+  - [optional] timezone offset in minutes as signed integer.
+    If this field is 0 then it's not saved.
+ */
+
+
+#define check_secs(secs)                                \
+	assert((int64_t)(secs) <= MAX_EPOCH_SECS_VALUE);\
+	assert((int64_t)(secs) >= MIN_EPOCH_SECS_VALUE);
+
+#define check_nanosecs(nsec)      assert((nsec) < 1000000000);
+
+#define check_tz_offset(offset)       \
+	assert((offset) <= (12 * 60));\
+	assert((offset) >= (-12 * 60));
+
+static inline uint32_t
+mp_sizeof_datetime_raw(const struct datetime *date)
+{
+	uint32_t sz = 0;
+	if (date->epoch != 0 || date->tzoffset != 0 ||
+	    date->tzindex != 0 || date->nsec != 0) {
+		check_secs(date->epoch);
+		sz += mp_sizeof_xint(date->epoch);
+	}
+	if (date->tzoffset != 0 || date->tzindex != 0 || date->nsec != 0) {
+		check_tz_offset(date->tzoffset);
+		sz += mp_sizeof_xint(date->tzoffset);
+	}
+	if (date->tzindex != 0 || date->nsec != 0) {
+		sz += mp_sizeof_xint(date->tzindex);
+	}
+	if (date->nsec != 0) {
+		check_nanosecs(date->nsec);
+		sz += mp_sizeof_xint(date->nsec);
+	}
+	return sz;
+}
+
+uint32_t
+mp_sizeof_datetime(const struct datetime *date)
+{
+	return mp_sizeof_ext(mp_sizeof_datetime_raw(date));
+}
+
+struct datetime *
+tnt_datetime_unpack(const char **data, uint32_t len, struct datetime *date)
+{
+	const char *svp = *data;
+	memset(date, 0, sizeof(*date));
+
+	if (len <= 0)
+		return date;
+
+	int64_t seconds = mp_decode_xint(data);
+	check_secs(seconds);
+	date->epoch = seconds;
+	len -= *data - svp;
+
+	if (len <= 0)
+		return date;
+
+	svp = *data;
+	int64_t tzoffset = mp_decode_xint(data);
+	check_tz_offset(tzoffset);
+	date->tzoffset = tzoffset;
+	len -= *data - svp;
+
+	if (len <= 0)
+		return date;
+
+	svp = *data;
+	int64_t tzindex = mp_decode_xint(data);
+	// FIXME - check_tzindex
+	date->tzoffset = tzindex;
+	len -= *data - svp;
+
+	if (len <= 0)
+		return date;
+
+	uint64_t nanoseconds = mp_decode_uint(data);
+	check_nanosecs(nanoseconds);
+	date->nsec = nanoseconds;
+
+	return date;
+}
+
+struct datetime *
+tnt_mp_decode_datetime(const char **data, struct datetime *date)
+{
+	if (mp_typeof(**data) != MP_EXT)
+		return NULL;
+
+	const char *svp = *data;
+	int8_t type;
+	uint32_t len = mp_decode_extl(data, &type);
+
+	if (type != MP_DATETIME) {
+		*data = svp;
+		return NULL;
+	}
+	return tnt_datetime_unpack(data, len, date);
```

Comment:
Here we need to check that `tnt_datetime_unpack` return NULL and restore *data to svp in this case as i think. See `mp_decode_uuid`.

Reply:
Oh, I see, here is the corrected caller side:
```diff
--- a/src/lib/core/mp_datetime.c
+++ b/src/lib/core/mp_datetime.c
@@ -91,11 +91,12 @@ tnt_mp_decode_datetime(const char **data, struct datetime *date)
 	int8_t type;
 	uint32_t len = mp_decode_extl(data, &type);
 
-	if (type != MP_DATETIME) {
+	if (type != MP_DATETIME ||
+	    tnt_datetime_unpack(data, len, date) == NULL) {
 		*data = svp;
 		return NULL;
 	}
-	return tnt_datetime_unpack(data, len, date);
+	return date;
 }
 
```